subprojects/core/src/main/java/org/gradle/api/internal/provider/GradlePropertiesCredentialsProviderFactory.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.provider;
+
+import org.gradle.api.Task;
+import org.gradle.api.credentials.PasswordCredentials;
+import org.gradle.api.execution.TaskExecutionGraph;
+import org.gradle.api.internal.properties.GradleProperties;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.TaskInputs;
+import org.gradle.internal.Describables;
+import org.gradle.internal.DisplayName;
+import org.gradle.internal.credentials.DefaultPasswordCredentials;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class GradlePropertiesCredentialsProviderFactory implements CredentialsProviderFactory {
+
+    private final GradleProperties gradleProperties;
+    private final TaskExecutionGraph taskExecutionGraph;
+
+    private boolean credentialsBeingEvaluated = false;
+
+    private final Map<String, PasswordCredentialsProvider> passwordProviders = new HashMap<>();
+
+    public GradlePropertiesCredentialsProviderFactory(GradleProperties gradleProperties, TaskExecutionGraph taskExecutionGraph) {
+        this.gradleProperties = gradleProperties;
+        this.taskExecutionGraph = taskExecutionGraph;
+    }
+
+    @Override
+    public Provider<PasswordCredentials> usernameAndPassword(String identity) {
+        validateIdentity(identity);
+        registerCredentialsEvaluator();
+        return passwordProviders.computeIfAbsent(identity, PasswordCredentialsProvider::new);
+    }
+
+    private void registerCredentialsEvaluator() {
+        if (!credentialsBeingEvaluated) {
+            taskExecutionGraph.whenReady(graph -> {
+                graph.getAllTasks().stream().map(Task::getInputs).forEach(TaskInputs::getProperties);
+                passwordProviders.values().stream().filter(p -> p.valueRequested).forEach(AbstractMinimalProvider::get);
+            });
```

Comment:
This is one way to validate credential presence before execution only when it is known that they will be needed and fail fast in such cases.
One (and probably major) downside is that it requires to traverse the task graph and evaluate all the declared inputs.